### PR TITLE
user authentication and tests

### PR DESCRIPTION
**Pull request by caustin-usgs**
@dhill-usgs @klsmith-usgs please take a look when you have time

### CODE REVIEW

**Comment by davidvhill** [comment]
Please create a single module that holds all constants, such as email_reg = '(\w+[.|\w])*@(\w+[.])*([a-zA-Z]*$)' in ordering_providers.py or file paths, etc.  Look up these vals from everywhere else.

**Comment by davidvhill** [comment]
wherever possible, try to use the dictcursor for postgres rather than index access.

**Comment by davidvhill** [comment]
nice work on the transport layer, keeping it clean.  Take care that User is not used in the api (if it needs to be shared, then rehost to a location where api does not import transport or you'll wind up in circular deps hell

**Comment by davidvhill** [comment]
you can nuke the is_empty and not_empty methods.  Iterators can be checked like so:
``` a_list = []
     if a_list:
         # evaluates to False
         pass

    a_list = [1,2,3]
    if a_list:
        # evaluates to True
        pass```

**Comment by davidvhill** [comment]
sensor.py:  products = "source_metadata l1 toa bt cloud sr lst swe sr_ndvi " \
               " sr_evi sr_savi sr_msavi sr_ndmi sr_nbr sr_nbr2 stats".split()

Could just do:
```
     products = ['source_metadata', 'l1',  'toa', 'bt', 'cloud', 'sr', 'lst', 'swe', 'sr_ndvi',
               'sr_evi', 'sr_savi', 'sr_msavi', 'sr_ndmi', 'sr_nbr', 'sr_nbr2', 'stats']
```

Or better yet, tuples are faster and immutable:
```
products = ('source_metadata', 'l1',  'toa', 'bt', 'cloud', 'sr', 'lst', 'swe', 'sr_ndvi',
               'sr_evi', 'sr_savi', 'sr_msavi', 'sr_ndmi', 'sr_nbr', 'sr_nbr2', 'stats')
```

**Comment by davidvhill** [comment]
domain.py

order.py was something i was considering, generating domain objects for the big nouns in the system and referencing that info throughout the codebase that way.  Totally up to you guys if you want to do that or not but it 'could' potentially make cleaner code.  (if you get into a place where you see the code repeatedly parsing, validating, checking, finding, then could refactor at that point.  Same reasoning behind creation of sensor.py)

**Comment by davidvhill** [comment]
nice job keeping version0.py clean.

**Comment by davidvhill** [comment]
highly recommended that you make each of your providers extend an abstract base class which contains the method signatures that need to be implemented.  That way it's clear what functionality is required for an alternate provider implementation (or Mock test object)

**Comment by davidvhill** [comment]
def fetch_order in ordering_provider.py

```
  sql = "select * from ordering_order where orderid = %s;"
        out_dict = {}
        opts_dict = {}
        scrub_keys = ['initial_email_sent', 'completion_email_sent', 'id', 'user_id',
			'ee_order_id', 'email']
```

Would be _way_ better to modify the query and only return the values necessary rather than pulling everything and filtering client side.

File: api-dev-uwsgi.ini [hunk 1]
```diff
@@ -6,17 +6,14 @@ callable=app
 master=True
 vacuum=True
 max-requests=5000
+DEBUG = True
```
Comment (caustin-usgs) [inline api-dev-uwsgi.ini]: need to sort out how to toggle this value based on where the app is running

File: api/lta.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+        Return:
+        The EE username
+        '''
+
+        return self.client.service.getUserName(contactid)
+
+
+
+
```
Comment (caustin-usgs) [inline api/lta.py]: @davidvhill are we going to need these classes ? OrderUpdateServiceClient and OrderDeliveryServiceClient

**Comment by klsmith-usgs** [comment]
Similar to how you have the API ABC setup? With the actual logic coming in
when it is instantiated, obscured a bit? This way there is a very clear way
of interacting with the provider as well (not just what is required), with
expected input and output values/types.

On Thu, Jan 28, 2016 at 2:19 PM, David V. Hill <notifications@github.com>
wrote:

> highly recommended that you make each of your providers extend an abstract
> base class which contains the method signatures that need to be
> implemented. That way it's clear what functionality is required for an
> alternate provider implementation (or Mock test object)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGS-EROS/espa-api/pull/3#issuecomment-176388414>.
>



-- 
Kelcy Smith
Software Engineer
Stinger Ghaffarian Technologies (SGT, Inc.)
Contractor to the U.S. Geological Survey (USGS)
Earth Resources Observation and Science (EROS) Center
47914 252nd St
Sioux Falls, South Dakota 57198
Office: (605) 594-2836
Fax: (605) 594-6529
Email: klsmith@usgs.gov

File: api/lta.py (hunk 2 above)
Comment (davidvhill) [inline api/lta.py]: We will most definitely need them

File: api-dev-uwsgi.ini (hunk 1 above)
Comment (davidvhill) [inline api-dev-uwsgi.ini]: That's handled in the current system.  Remind me to show you how

Comment (caustin-usgs) [inline api-dev-uwsgi.ini]: https://github.com/USGS-EROS/espa-api/commit/11a7ffbaa6113df56f7fc9f04f0981941e06e4bf
i thought we could handle this slightly different then is done in espa-web.  there's a new config module in api/domain. that determines the current operating environment based on the properties set in the .cfgnfo file for the user running the app.
when in 'dev' mode, turn on Flask debugging. works when running on uwsgi as well. also allows turning on transport layer debugging by setting an environment variable, ESPA_DEBUG=True

**Pull request merged**